### ISSUE TITLE
Fix composite-metazoan pipeline 

### ISSUE THREAD

**Issue by shawntanzk** [obophenotype/uberon#2588]
1) Look at https://github.com/obophenotype/uberon/issues/2584 -> how did CL relationship go missing? (CL variant that is imported in this process that is probably not the right one? or CL is not imported?) 
2) https://github.com/obophenotype/uberon/issues/1466
3) ?

**Comment by gouttegd** [comment]
From the Uberon call of October 24th, 2022: The whole point of the `composite-*` products is precisely to have the kind of “unfolded” structure that `owltools --merge-species-ontology` generates, so this should not be changed.

-> Keep the `owltools`-dependent pipeline for now, and start thinking whether the `--merge-species-ontology` feature should be ported/re-implemented elsewhere.

**Comment by aschroed** [comment]
Is there anywhere that a recent version of composite-metazoan.owl is available for download and use?

**Comment by anitacaron** [comment]
Hi @aschroed, the `composite-metazoan.owl` is now available in the [release assets](https://github.com/obophenotype/uberon/releases).

**Comment by gouttegd** [comment]
I propose to use this ticket to centralise all discussions about the overhauling of the `composite-metazoan` (hereafter CM) pipeline.

Here’s a list of all the issues affecting this product. I strongly suspect most if not all are directly caused by the complexity of the CM pipeline (and the lack of love shown to that pipeline over the last few years), and therefore most if not all of them could be fixed by streamlining the pipeline (one can always hope, at least).

- [ ] #1466
- [ ] #1813
- [ ] #1952
- [ ] #1957
- [ ] #1958
- [ ] #1959
- [ ] #2032
- [ ] #2584
- [ ] #2782

Regarding #1952, the unsats may be caused by the actual contents of the merged ontologies and as such may not be fixed by a new pipeline, _but_ a cleaner, faster pipeline may make testing for unsats easier and help detecting and fixing those in the future.

**Comment by gouttegd** [comment]
Here’s a quick overview of the current CM pipeline.

## Step 1: $TMPDIR/merged-composite-metazoan.owl

This first step is merely a big merge of a bunch of files:

* `$TMPDIR/ext-weak.owl` (mostly Uberon itself, minus some disjointness axioms and equivalent-to-nothing axioms);
* all the cross-ontology bridges (e.g. `uberon-to-fbbt-bridge.owl`, etc.);
* `imports/local-*.owl` (where `*` = `fbbt`, `wbbt`, `zfa`, etc.)
  * these files are generated from the corresponding mirrors, in what seems to be a _manual_ step that is not part of any pipeline (they were last updated in 2021);
  * according to a comment in the Makefile, a `local-something` file is a version of the `something` mirror “with all the wrangling needed to massage it it into the right form”;
  * whether that “wrangling” is still needed nowadays is unclear (a later comment mentions that “many ontologies have fixed their legacy properties”, hinting that this step _may_ no longer be needed;
  * some `local-*` files are generated by dedicated custom rules (again, whether they are still really needed or not is unclear);
* `$TMPDIR/allen-*.obo` (a bunch of files generated from JSON files downloaded from api.brain-map.org)
* `uberon.owl` (this seems redundant with `ext-weak.owl` to me);
* all the `bridges/collected-*.owl` files
  * those are small files that merely bind together several ontologies via `imports` declaration;
  * for example, `collected-mouse.owl` binds `uberon` and `emapa` by importing `uberon.owl`, `emapa.owl`, and the corresponding bridges `uberon-bridge-to-emapa.owl` and `cl-bridge-to-emapa.owl`.

## Step 2: $TMPDIR/stripped-composite-metazoan.owl

This file is built from the previous one (`merged-composite-metazoan.owl`) through a single ROBOT-based step to remove some unsatisfiable classes that are listed in the `unsats.txt` file (last updated in 2021).


## Step 3: $TMPDIR/unreasoned-composite-metazoan.owl

This file is produced from the previous one by a OWLTOOLS-based reasoning step. That step is critically dependent on OWLTOOLS as it requires two commands that have no equivalents AFAIK in other tools (e.g. ROBOT):

* `--merge-species-ontology`, which performs “unfolding” as described in [OWLTOOLS’s source code](https://github.com/owlcollab/owltools/blob/master/OWLTools-Core/src/main/java/owltools/mooncat/SpeciesMergeUtil.java);
* `--merge-equivalence-sets`, which merges equivalent classes with control of which ontology “wins” when a class in one ontology is equivalent to a class in another ontology (see [OWLTOOLS’s source code](https://github.com/owlcollab/owltools/blob/master/OWLTools-Core/src/main/java/owltools/mooncat/EquivalenceSetMergeUtil.java)).


## Step 4: composite-metazoan.owl

Lastly, the final `composite-metazoan.owl` product is generated from `unreasoned-composite-metazoan.owl` by what seems to be a fairly standard ROBOT-based reasoning step.

**Comment by gouttegd** [comment]
For reference and discussion, here are some of the changes I plan to introduce to fix/overhaul the composite pipeline.

## Make sure CL is merged in every composite
Currently, CL is _not_ explicitly merged in the composite ontologies – it is only merged through Uberon’s CL import. As a result, `composite-metazoan` only contains about 50% of CL’s terms. Several projects are using `composite-metazoan` to annotate single-cell RNA-sequencing data, so it should include all of CL. All the CL bridges with taxon-specific ontologies are already generated in Uberon, so excluding CL from `composite-metazoan` makes little sense.

## Use the Uberon.Makefile as sole source of truth for the composition of the collected ontologies
With the current pipeline, the various components that make up a composite ontology (including `composite-metazoan`) are decided in at least three different places:

1) in the `src/ontology/bridge/collected-NAME.owl` file, which contains import declarations for each component, e.g.:

```xml
    <owl:imports rdf:resource="http://purl.obolibrary.org/obo/uberon/bridge/collected-vertebrate.owl"/>
    <owl:imports rdf:resource="http://purl.obolibrary.org/obo/uberon/bridge/uberon-bridge-to-fbbt.owl"/>
    <owl:imports rdf:resource="http://purl.obolibrary.org/obo/uberon/bridge/uberon-bridge-to-wbbt.owl"/>
    <owl:imports rdf:resource="http://purl.obolibrary.org/obo/uberon/bridge/uberon-bridge-to-fbdv.owl"/>
    <owl:imports rdf:resource="http://purl.obolibrary.org/obo/uberon/bridge/uberon-bridge-to-wbls.owl"/>
    <owl:imports rdf:resource="http://purl.obolibrary.org/obo/uberon/bridge/cl-bridge-to-fbbt.owl"/>
    <owl:imports rdf:resource="http://purl.obolibrary.org/obo/uberon/bridge/cl-bridge-to-wbbt.owl"/>
    <owl:imports rdf:resource="http://purl.obolibrary.org/obo/uberon/ssso-merged-uberon.owl"/>
    <owl:imports rdf:resource="http://purl.obolibrary.org/obo/fbbt.owl"/>
    <owl:imports rdf:resource="http://purl.obolibrary.org/obo/fbdv.owl"/>
    <owl:imports rdf:resource="http://purl.obolibrary.org/obo/wbbt.owl"/>
    <owl:imports rdf:resource="http://purl.obolibrary.org/obo/wbls.owl"/>
    <owl:imports rdf:resource="http://purl.obolibrary.org/obo/ceph.owl"/>
    <owl:imports rdf:resource="http://purl.obolibrary.org/obo/cteno.owl"/>
    <owl:imports rdf:resource="http://purl.obolibrary.org/obo/poro.owl"/>
```

2) in the `src/ontology/catalog-v001.xml`, which redirects the ontology IRIs to local resources, e.g.:

```xml
      <uri name="http://purl.obolibrary.org/obo/uberon/ssso-merged-uberon.owl" uri="tmp/developmental-stage-ontologies/src/ssso-merged.obo"/>
      <uri name="http://purl.obolibrary.org/obo/uberon/bridge/collected-vertebrate.owl" uri="bridge/collected-vertebrate.owl"/>
      <uri name="http://purl.obolibrary.org/obo/uberon/bridge/collected-tetrapod.owl" uri="bridge/collected-tetrapod.owl"/>
      <uri name="http://purl.obolibrary.org/obo/uberon/bridge/collected-teleost.owl" uri="bridge/collected-teleost.owl"/>
      <uri name="http://purl.obolibrary.org/obo/uberon/bridge/collected-amniote.owl" uri="bridge/collected-amniote.owl"/>
      <uri name="http://purl.obolibrary.org/obo/uberon/bridge/collected-mammal.owl" uri="bridge/collected-mammal.owl"/>
```

3) in the `src/ontology/uberon.Makefile`, which does amusing things like this:

```sh
owltools --catalog-xml catalog-v001.xml \
         --map-ontology-iri $URIBASE/uberon.owl $TMPDIR/ext-weak.owl \
         --map-ontology-iri $URIBASE/fma.owl $COMPONENTSDIR/null.owl \
         --map-ontology-iri $URIBASE/uberon/bridge/uberon-bridge-to-fma.owl $COMPONENTSDIR/null.owl
```

All of this makes it needlessly complicated to figure out what exactly goes in each composite product.

Therefore, I plan to get rid of both the import declarations in the `src/ontology/bridge/collected-*.owl` files and the use of the catalog. All the components that make up a given composite ontology will be listed in `uberon.Makefile` as pre-requisites of that composite ontology, and will be “manually” merged together by `robot merge`. This will make for a somewhat long section in the Makefile, but at least everything will be at the same place.

## Publish `collected-metazoan` (at least) as a release artefact
`collected-metazoan` is an intermediate product towards `composite-metazoan`. It’s basically `composite-metazoan` before any reasoning operation is performed. I think it’s a useful product in its own right, especially for annotation purposes, so it should be published as a release artefact.

## Commit copies of all external resources, and only refresh them explicitly
Several components of `composite-metazoan` are automatically downloaded every time the pipeline is run. I want to generalise the approach already used in the bridge pipeline, in which all external resources are committed to the repository as local copies, and only refreshed when 1) we run the pipeline under `IMP=true` (which we do not as part of the QC or release process) or 2) we explicitly request a refresh (`make refresh-external-resources`).

**Comment by gouttegd** [comment]
I also plan to deprecate the following `composite-` products:

* `composite-mouse-plus-stage`: this relies on EMAP, which is deprecated and whose bridge has already been retired;
* `composite-vhog`: VHOG is deprecated and its bridge has already been retired;
* `composite-teleost`: this is in effect the same product as `composite-zebrafish`;
* `composite-fma`: we have been explicitly excluding FMA from `composite-metazoan.owl` for a while, so I doubt there is an interest specifically for this one.

There are also two other products that are “larger” than `composite-metazoan` (i.e. they are made of `composite-metazoan` plus some additional ontologies):

* `composite-opisthokont` (= `collected-metazoan` + FAO);
* `composite-eukaryote` (= `collected-opisthokont + PO + DDANAT)

The mere fact that we have always only ever talked about `composite-metazoan` and not those two, not to mention the fact that they have never been published as proper release artefact, strongly suggests that there is little interest in them and I plan to retire them as well.

**Comment by gouttegd** [comment]
Fixed by #3129
